File: src/main/java/no/nav/aareg/teknisk/historikk/filter/RequestHeaderFilter.java
package no.nav.aareg.teknisk.historikk.filter;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.UUID;

@Component
public class RequestHeaderFilter extends OncePerRequestFilter {

    public static final String KORRELASJONSID_HEADER = "correlation-id";

    @Override
    protected void doFilterInternal(HttpServletRequest req, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        try {
            var korrelasjonsid = req.getHeader(KORRELASJONSID_HEADER);
            if (korrelasjonsid == null || korrelasjonsid.isBlank()) {
                korrelasjonsid = UUID.randomUUID().toString();
            }

            req.setAttribute(KORRELASJONSID_HEADER, korrelasjonsid);
            response.setHeader(KORRELASJONSID_HEADER, korrelasjonsid);

            filterChain.doFilter(req, response);
        } finally {
            // no-op
        }
    }
}

File: src/main/kotlin/no/nav/aareg/teknisk_historikk/RequestHeaderFilter.kt
package no.nav.aareg.teknisk_historikk

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.util.UUID.randomUUID

const val KORRELASJONSID_HEADER = "correlation-id"

@Component
class RequestHeaderFilter : OncePerRequestFilter() {
    override fun doFilterInternal(req: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        try {
            val korrelasjonsid = req.getHeader(KORRELASJONSID_HEADER) ?: randomUUID().toString()

            MDC.put(KORRELASJONSID_HEADER, korrelasjonsid)
            req.setAttribute(KORRELASJONSID_HEADER, korrelasjonsid)
            response.addHeader(KORRELASJONSID_HEADER, korrelasjonsid)

            filterChain.doFilter(req, response)
        } finally {
            MDC.remove(KORRELASJONSID_HEADER)
        }
    }
}